### PR TITLE
Add gradient fusion method for multi-partner-learning

### PR DESCRIPTION
**Pull request by arthurPignet**
This PR follows the [PR271](https://github.com/SubstraFoundation/distributed-learning-contributivity/pull/271)

### CODE REVIEW

File: mplc/multi_partner_learning.py [hunk 1]
```diff
@@ -603,12 +604,72 @@ def fit_minibatch(self):
...
+            # Evaluate and store accuracy of mini-batch start model
+            partner.model_weights = self.model_weights
+        self.eval_and_log_model_val_perf()
+
+        # Iterate over partners for training each individual model
+        for partner_index, partner in enumerate(self.partners_list):
+            with tf.GradientTape() as tape:
+                loss = self.model.loss(partner.minibatched_y_train[self.minibatch_index],
+                                       self.model(partner.minibatched_x_train[self.minibatch_index]))
+            partner.grads = tape.gradient(loss, self.model.trainable_weights)
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning.py:643]: So in this case it's like `gradient_updates_per_pass_count=1` ?

File: mplc/multi_partner_learning.py [hunk 2]
```diff
@@ -603,12 +604,72 @@ def fit_minibatch(self):
...
 # Supported multi-partner learning approaches
 
 MULTI_PARTNER_LEARNING_APPROACHES = {
     "fedavg": FederatedAverageLearning,
     "seq-pure": SequentialLearning,
     "seq-with-final-agg": SequentialWithFinalAggLearning,
     "seqavg": SequentialAverageLearning,
-    "lflip": MplSModel
+    "lflip": MplSModel,
+    'grads-fusion': GradientFusion
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning.py:674]: GradientsFusion with an s

File: mplc/multi_partner_learning.py [hunk 3]
```diff
@@ -603,12 +604,72 @@ def fit_minibatch(self):
...
+
+        for partner in self.partners_list:
+            # Evaluate and store accuracy of mini-batch start model
+            partner.model_weights = self.model_weights
+        self.eval_and_log_model_val_perf()
+
+        # Iterate over partners for training each individual model
+        for partner_index, partner in enumerate(self.partners_list):
+            with tf.GradientTape() as tape:
+                loss = self.model.loss(partner.minibatched_y_train[self.minibatch_index],
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning.py:641]: this fails on titanic dataset.
We could either make it work on titanic or write somewhere that this method only works with neural networks.

**Review by RomainGoussault — CHANGES_REQUESTED** [review]

File: mplc/multi_partner_learning.py (hunk 1 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning.py:643]: yes

File: mplc/multi_partner_learning.py (hunk 3 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning.py:641]: I specified it in the doc

File: mplc/multi_partner_learning.py (hunk 2 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning.py:674]: changed to GradientsAverage

**Review by RomainGoussault — APPROVED** [review]
Looks good to me!

**Review by bowni — APPROVED** [review]

**Review by natct10 — APPROVED** [review]
Neat!
